File: parser/parser.go
package parser

import (
	"fmt"
	"github.com/ippan/clover/ast"
	"github.com/ippan/clover/lexer"
	"github.com/ippan/clover/token"
	"strconv"
)

const (
	_ int = iota
	LOWEST
	ASSIGN
	BOOLEAN
	EQUALS
	LESSGREATER
	SUM
	PRODUCT
	PREFIX
	CALL
	DOT
)

var precedences = map[token.TokenType]int{
	token.ASSIGN:           ASSIGN,
	token.PLUS_ASSIGN:      ASSIGN,
	token.MINUS_ASSIGN:     ASSIGN,
	token.STAR_ASSIGN:      ASSIGN,
	token.SLASH_ASSIGN:     ASSIGN,
	token.AND:              BOOLEAN,
	token.OR:               BOOLEAN,
	token.EQUAL:            EQUALS,
	token.NOT_EQUAL:        EQUALS,
	token.LESS:             LESSGREATER,
	token.GREATER:          LESSGREATER,
	token.LESS_EQUAL:       LESSGREATER,
	token.GREATER_EQUAL:    LESSGREATER,
	token.PLUS:             SUM,
	token.MINUS:            SUM,
	token.STAR:             PRODUCT,
	token.SLASH:            PRODUCT,
	token.BIT_AND:          PRODUCT,
	token.BIT_OR:           PRODUCT,
	token.DOT:              DOT,
	token.LEFT_PARENTHESES: CALL,
}

type prefixFunction func() ast.Expression
type infixFunction func(expression ast.Expression) ast.Expression

type Parser struct {
	l      *lexer.Lexer
	errors []string

	currentToken token.Token
	peekToken    token.Token

	currentTokenLine int
	peekTokenLine    int

	prefixFunctions map[token.TokenType]prefixFunction
	infixFunctions  map[token.TokenType]infixFunction
}

func New(l *lexer.Lexer) *Parser {
	p := &Parser{l: l}

	p.prefixFunctions = make(map[token.TokenType]prefixFunction)

	p.registerPrefixFunction(token.IDENTIFIER, p.parseIdentifier)
	p.registerPrefixFunction(token.BASE, p.parseBaseLiteral)
	p.registerPrefixFunction(token.THIS, p.parseThisLiteral)
	p.registerPrefixFunction(token.INTEGER, p.parseIntegerLiteral)
	p.registerPrefixFunction(token.FLOAT, p.parseFloatLiteral)
	p.registerPrefixFunction(token.TRUE, p.parseBooleanLiteral)
	p.registerPrefixFunction(token.FALSE, p.parseBooleanLiteral)
	p.registerPrefixFunction(token.STRING, p.parseStringLiteral)
	p.registerPrefixFunction(token.NULL, p.parseNullLiteral)
	p.registerPrefixFunction(token.MINUS, p.parsePrefixExpression)
	p.registerPrefixFunction(token.NOT, p.parsePrefixExpression)
	p.registerPrefixFunction(token.LEFT_PARENTHESES, p.parseGroupExpression)
	p.registerPrefixFunction(token.IF, p.parseIfExpression)
	p.registerPrefixFunction(token.WHILE, p.parseWhileExpression)
	p.registerPrefixFunction(token.FUNCTION, p.parseFunctionExpression)
	p.registerPrefixFunction(token.CLASS, p.parseClassExpression)

	p.infixFunctions = make(map[token.TokenType]infixFunction)

	p.registerInfixFunction(token.ASSIGN, p.parseInfixExpression)
	p.registerInfixFunction(token.PLUS_ASSIGN, p.parseInfixExpression)
	p.registerInfixFunction(token.MINUS_ASSIGN, p.parseInfixExpression)
	p.registerInfixFunction(token.STAR_ASSIGN, p.parseInfixExpression)
	p.registerInfixFunction(token.SLASH_ASSIGN, p.parseInfixExpression)
	p.registerInfixFunction(token.AND, p.parseInfixExpression)
	p.registerInfixFunction(token.OR, p.parseInfixExpression)
	p.registerInfixFunction(token.EQUAL, p.parseInfixExpression)
	p.registerInfixFunction(token.NOT_EQUAL, p.parseInfixExpression)
	p.registerInfixFunction(token.LESS, p.parseInfixExpression)
	p.registerInfixFunction(token.GREATER, p.parseInfixExpression)
	p.registerInfixFunction(token.LESS_EQUAL, p.parseInfixExpression)
	p.registerInfixFunction(token.GREATER_EQUAL, p.parseInfixExpression)
	p.registerInfixFunction(token.PLUS, p.parseInfixExpression)
	p.registerInfixFunction(token.MINUS, p.parseInfixExpression)
	p.registerInfixFunction(token.STAR, p.parseInfixExpression)
	p.registerInfixFunction(token.SLASH, p.parseInfixExpression)
	p.registerInfixFunction(token.BIT_AND, p.parseInfixExpression)
	p.registerInfixFunction(token.BIT_OR, p.parseInfixExpression)
	p.registerInfixFunction(token.DOT, p.parseInfixExpression)
	p.registerInfixFunction(token.LEFT_PARENTHESES, p.parseCallExpression)

	// run twice to set current and peek token
	p.nextToken()
	p.nextToken()

	return p
}

func (p *Parser) Errors() []string {
	return p.errors
}

func (p *Parser) pushError(error string) {
	p.errors = append(p.errors, fmt.Sprintf("[line %d] %s", p.currentTokenLine, error))
}

func (p *Parser) nextToken() {
	p.currentToken = p.peekToken
	p.currentTokenLine = p.peekTokenLine

	p.peekToken = p.l.Lex()
	p.peekTokenLine = p.l.TokenLine()
}

func (p *Parser) currentTokenIs(t token.TokenType) bool {
	return p.currentToken.Type == t
}

func (p *Parser) peekTokenIs(t token.TokenType) bool {
	return p.peekToken.Type == t
}

func (p *Parser) expectToken(t token.Token, tt token.TokenType) bool {
	if t.Type == tt {
		return true
	}
	p.pushError(fmt.Sprintf("expected token type is %s, got %s", tt, t.Type))
	return false
}

func (p *Parser) expectPeek(t token.TokenType) bool {
	if p.expectToken(p.peekToken, t) {
		p.nextToken()
		return true
	}

	return false
}

func (p *Parser) currentPrecedence() int {
	if precedence, ok := precedences[p.currentToken.Type]; ok {
		return precedence
	}

	return LOWEST
}

func (p *Parser) peekPrecedence() int {
	if precedence, ok := precedences[p.peekToken.Type]; ok {
		return precedence
	}

	return LOWEST
}

func (p *Parser) registerPrefixFunction(tt token.TokenType, function prefixFunction) {
	p.prefixFunctions[tt] = function
}

func (p *Parser) registerInfixFunction(tt token.TokenType, function infixFunction) {
	p.infixFunctions[tt] = function
}

func (p *Parser) parseProgram(terminators []token.TokenType) *ast.Program {
	program := &ast.Program{Statements: []ast.Statement{}}

	for {
		for _, terminator := range terminators {
			if p.currentTokenIs(terminator) {
				return program
			}
		}

		statement := p.parseStatement()
		if statement != nil {
			program.Statements = append(program.Statements, statement)
		}
		p.nextToken()
	}
}

func (p *Parser) Parse() *ast.Program {
	return p.parseProgram([]token.TokenType{token.EOF})
}

func (p *Parser) parseExpressionStatement() *ast.ExpressionStatement {
	statement := &ast.ExpressionStatement{Token: p.currentToken}

	statement.Expression = p.parseExpression(LOWEST)

	return statement
}

func (p *Parser) parseReturnStatement() *ast.ReturnStatement {
	statement := &ast.ReturnStatement{Token: p.currentToken}

	// have return value?
	if (p.currentTokenLine == p.peekTokenLine) && !p.peekTokenIs(token.EOF) {
		p.nextToken()
		statement.ReturnValue = p.parseExpression(LOWEST)
	}

	return statement
}

func (p *Parser) parseStatement() ast.Statement {
	switch p.currentToken.Type {
	case token.RETURN:
		return p.parseReturnStatement()
	}

	return p.parseExpressionStatement()
}

func (p *Parser) parseExpression(precedence int) ast.Expression {
	prefix := p.prefixFunctions[p.currentToken.Type]

	if prefix == nil {
		p.pushError(fmt.Sprintf("prefix function for %s does not exists", p.currentToken.Literal))
		return nil
	}

	leftExpression := prefix()

	for !p.peekTokenIs(token.EOF) && precedence < p.peekPrecedence() {
		infix, ok := p.infixFunctions[p.peekToken.Type]
		if !ok {
			return leftExpression
		}

		p.nextToken()

		leftExpression = infix(leftExpression)
	}

	return leftExpression
}

func (p *Parser) parseIdentifier() ast.Expression {
	return &ast.Identifier{Token: p.currentToken, Value: p.currentToken.Literal}
}

func (p *Parser) parseBaseLiteral() ast.Expression {
	return &ast.BaseLiteral{Token: p.currentToken}
}

func (p *Parser) parseThisLiteral() ast.Expression {
	return &ast.ThisLiteral{Token: p.currentToken}
}

func (p *Parser) parseIntegerLiteral() ast.Expression {
	it := &ast.IntegerLiteral{Token: p.currentToken}

	value, error := strconv.ParseInt(p.currentToken.Literal, 0, 64)

	if error != nil {
		p.pushError(fmt.Sprintf("%q is not an integer", p.currentToken.Literal))
	}

	it.Value = value

	return it
}

func (p *Parser) parseFloatLiteral() ast.Expression {
	ft := &ast.FloatLiteral{Token: p.currentToken}

	value, error := strconv.ParseFloat(p.currentToken.Literal, 64)

	if error != nil {
		p.pushError(fmt.Sprintf("%q is not a float", p.currentToken.Literal))
	}

	ft.Value = value

	return ft
}

func (p *Parser) parseBooleanLiteral() ast.Expression {
	return &ast.BooleanLiteral{Token: p.currentToken, Value: p.currentTokenIs(token.TRUE)}
}

func (p *Parser) parseStringLiteral() ast.Expression {
	return &ast.StringLiteral{Token: p.currentToken}
}

func (p *Parser) parseNullLiteral() ast.Expression {
	return &ast.NullLiteral{Token: p.currentToken}
}

func (p *Parser) parsePrefixExpression() ast.Expression {
	pe := &ast.PrefixExpression{Token: p.currentToken}

	p.nextToken()

	pe.Right = p.parseExpression(PREFIX)

	return pe
}

func (p *Parser) parseInfixExpression(left ast.Expression) ast.Expression {
	ie := &ast.InfixExpression{Left: left, Token: p.currentToken}

	precedence := p.currentPrecedence()

	p.nextToken()

	ie.Right = p.parseExpression(precedence)

	return ie
}

func (p *Parser) parseGroupExpression() ast.Expression {
	p.nextToken()

	expression := p.parseExpression(LOWEST)

	if !p.expectPeek(token.RIGHT_PARENTHESES) {
		return nil
	}

	return expression
}

func (p *Parser) parseIfExpression() ast.Expression {
	ie := &ast.IfExpression{Token: p.currentToken}

	if !p.expectPeek(token.LEFT_PARENTHESES) {
		return nil
	}

	p.nextToken()
	ie.Condition = p.parseExpression(LOWEST)

	if !p.expectPeek(token.RIGHT_PARENTHESES) {
		return nil
	}

	p.nextToken()

	ie.TruePart = p.parseProgram([]token.TokenType{token.ELSE, token.END})

	if p.currentTokenIs(token.ELSE) {
		p.nextToken()
		ie.FalsePart = p.parseProgram([]token.TokenType{token.END})
	}

	if !p.expectToken(p.currentToken, token.END) {
		return nil
	}

	return ie
}

func (p *Parser) parseWhileExpression() ast.Expression {
	we := &ast.WhileExpression{Token: p.currentToken}

	if !p.expectPeek(token.LEFT_PARENTHESES) {
		return nil
	}

	p.nextToken()
	we.Condition = p.parseExpression(LOWEST)

	if !p.expectPeek(token.RIGHT_PARENTHESES) {
		return nil
	}

	p.nextToken()

	we.Body = p.parseProgram([]token.TokenType{token.END})

	if !p.expectToken(p.currentToken, token.END) {
		return nil
	}

	return we
}

func (p *Parser) parseParameters(separator token.TokenType, assignTokenType token.TokenType, terminator token.TokenType) []*ast.Parameter {
	var parameters []*ast.Parameter

	for !p.currentTokenIs(terminator) {

		if !p.expectToken(p.currentToken, token.IDENTIFIER) {
			return nil
		}

		parameter := &ast.Parameter{Name: p.parseIdentifier().(*ast.Identifier)}

		if p.peekTokenIs(assignTokenType) {
			p.nextToken()
			p.nextToken()
			parameter.Value = p.parseExpression(LOWEST)
		}

		parameters = append(parameters, parameter)

		if separator != token.EMPTY && !p.peekTokenIs(terminator) {
			if !p.expectPeek(separator) {
				return nil
			}
		}

		p.nextToken()
	}

	return parameters
}

func (p *Parser) parseFunctionExpression() ast.Expression {
	fe := &ast.FunctionExpression{Token: p.currentToken}

	if !p.expectPeek(token.LEFT_PARENTHESES) {
		return nil
	}

	p.nextToken()

	fe.Parameters = p.parseParameters(token.COMMA, token.ASSIGN, token.RIGHT_PARENTHESES)

	p.nextToken()

	fe.Body = p.parseProgram([]token.TokenType{token.END})

	if !p.expectToken(p.currentToken, token.END) {
		return nil
	}

	return fe
}

func (p *Parser) parseCallExpression(function ast.Expression) ast.Expression {
	ce := &ast.CallExpression{Function: function}

	if !p.expectToken(p.currentToken, token.LEFT_PARENTHESES) {
		return nil
	}

	p.nextToken()

	var arguments []ast.Expression

	for !p.currentTokenIs(token.RIGHT_PARENTHESES) {
		expression := p.parseExpression(LOWEST)
		arguments = append(arguments, expression)

		if !p.peekTokenIs(token.RIGHT_PARENTHESES) {
			if !p.expectPeek(token.COMMA) {
				return nil
			}
		}

		p.nextToken()
	}

	ce.Arguments = arguments

	return ce
}

func (p *Parser) parseClassExpression() ast.Expression {
	ce := &ast.ClassExpression{Token: p.currentToken}

	if p.peekTokenIs(token.EXTENDS) {
		p.nextToken()
		p.nextToken()
		ce.Parent = p.parseExpression(LOWEST)
	}

	p.nextToken()

	ce.Body = p.parseParameters(token.EMPTY, token.ASSIGN, token.END)

	if !p.expectToken(p.currentToken, token.END) {
		return nil
	}

	return ce
}

File: Clover/Parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Transactions;
using Clover.Ast;

namespace Clover
{
    
    public interface FileLoader
    {
        String LoadFile(String base_filename, String filename);
    }

    public class DefaultFileLoader : FileLoader
    {
        public string LoadFile(string base_filename, string filename)
        {
            if (String.IsNullOrEmpty(base_filename))
                return File.ReadAllText(filename);

            return string.Empty;
        }
    }

    public class Parser
    {
        private delegate Expression PrefixFunction();
        private delegate Expression InfixFunction(Expression expression);
        
        private enum SymbolPriority
        {
            Lowest = 0,
            Assign,
            Boolean,
            Equals,
            LessGreater,
            Sum,
            Product,
            Prefix,
            Call,
            InstanceGet
        }

        private static readonly Dictionary<Token, SymbolPriority> precedences = new Dictionary<Token, SymbolPriority>
        {
            { Token.Assign, SymbolPriority.Assign },
            { Token.PlusAssign, SymbolPriority.Assign },
            { Token.MinusAssign, SymbolPriority.Assign },
            { Token.StarAssign, SymbolPriority.Assign },
            { Token.SlashAssign, SymbolPriority.Assign },
            { Token.And, SymbolPriority.Boolean },
            { Token.Or, SymbolPriority.Boolean },
            { Token.Equal, SymbolPriority.Equals },
            { Token.NotEqual, SymbolPriority.Equals },
            { Token.Less, SymbolPriority.LessGreater },
            { Token.Greater, SymbolPriority.LessGreater },
            { Token.LessEqual, SymbolPriority.LessGreater },
            { Token.GreaterEqual, SymbolPriority.LessGreater },
            { Token.Plus, SymbolPriority.Sum },
            { Token.Minus, SymbolPriority.Sum },
            { Token.Star, SymbolPriority.Product },
            { Token.Slash, SymbolPriority.Product },
            { Token.BitAnd, SymbolPriority.Product },
            { Token.BitOr, SymbolPriority.Product },
            { Token.Dot, SymbolPriority.InstanceGet },
            { Token.LeftBracket, SymbolPriority.InstanceGet },
            { Token.LeftParentheses, SymbolPriority.Call }
        };
        
        private Lexer lexer;

        private readonly FileLoader file_loader;

        private readonly Dictionary<Token, PrefixFunction> prefix_functions = new Dictionary<Token, PrefixFunction>();
        private readonly Dictionary<Token, InfixFunction> infix_functions = new Dictionary<Token, InfixFunction>();

        private TokenData CurrentTokenData;
        private TokenData PeekTokenData;
        
        public Parser(FileLoader loader = null)
        {
            file_loader = loader ?? new DefaultFileLoader();

            prefix_functions[Token.Identifier] = ParseIdentifier;
            prefix_functions[Token.Base] = ParseKeywordLiteral;
            prefix_functions[Token.This] = ParseKeywordLiteral;
            prefix_functions[Token.Null] = ParseKeywordLiteral;
            prefix_functions[Token.String] = ParseStringLiteral;
            prefix_functions[Token.Integer] = ParseIntegerLiteral;
            prefix_functions[Token.Float] = ParseFloatLiteral;
            prefix_functions[Token.True] = ParseBooleanLiteral;
            prefix_functions[Token.False] = ParseBooleanLiteral;
            prefix_functions[Token.Minus] = ParsePrefixExpression;
            prefix_functions[Token.Not] = ParsePrefixExpression;
            prefix_functions[Token.LeftParentheses] = ParseGroupExpression;
            prefix_functions[Token.If] = ParseIfExpression;
            prefix_functions[Token.Local] = ParseLocalExpression;
            prefix_functions[Token.Function] = ParseFunctionExpression;
            prefix_functions[Token.Return] = ParseReturnExpression;
            prefix_functions[Token.LeftBracket] = ParseArrayExpression;

            infix_functions[Token.Assign] = ParseInfixExpression;
            infix_functions[Token.PlusAssign] = ParseInfixExpression;
            infix_functions[Token.MinusAssign] = ParseInfixExpression;
            infix_functions[Token.StarAssign] = ParseInfixExpression;
            infix_functions[Token.SlashAssign] = ParseInfixExpression;
            infix_functions[Token.And] = ParseInfixExpression;
            infix_functions[Token.Or] = ParseInfixExpression;
            infix_functions[Token.Equal] = ParseInfixExpression;
            infix_functions[Token.NotEqual] = ParseInfixExpression;
            infix_functions[Token.Less] = ParseInfixExpression;
            infix_functions[Token.Greater] = ParseInfixExpression;
            infix_functions[Token.LessEqual] = ParseInfixExpression;
            infix_functions[Token.GreaterEqual] = ParseInfixExpression;
            infix_functions[Token.Plus] = ParseInfixExpression;
            infix_functions[Token.Minus] = ParseInfixExpression;
            infix_functions[Token.Star] = ParseInfixExpression;
            infix_functions[Token.Slash] = ParseInfixExpression;
            infix_functions[Token.BitAnd] = ParseInfixExpression;
            infix_functions[Token.BitOr] = ParseInfixExpression;
            infix_functions[Token.Dot] = ParseInstanceGetExpression;
            infix_functions[Token.LeftBracket] = ParseInstanceGetExpression;
            infix_functions[Token.LeftParentheses] = ParseCallExpression;
        }

        private Expression ParseArrayExpression()
        {
            ArrayExpression array_expression = new ArrayExpression();
            
            NextToken();

            array_expression.Values = ParseCommaExpressions(Token.RightBracket);
            
            return array_expression;
        }

        private List<Expression> ParseCommaExpressions(Token end_token)
        {
            List<Expression> expressions = new List<Expression>();
            
            bool last_is_comma = false;
            
            while (!(CurrentTokenIs(end_token) || CurrentTokenIs(Token.Eof)))
            {
                expressions.Add(ParseExpression());
                
                NextToken();
                
                if (CurrentTokenIs(Token.Comma))
                {
                    last_is_comma = true;
                    NextToken();
                }
                else
                {
                    last_is_comma = false;
                }
            }

            if (last_is_comma)
            {
                // TODO : raise error
            }

            if (!CurrentTokenIs(end_token))
            {
                // TODO : raise error
            }
            
            return expressions;
        }

        private Expression ParseInstanceGetExpression(Expression instnace)
        {
            InstanceGetExpression instance_get_expression = new InstanceGetExpression { Instance = instnace };

            if (CurrentTokenIs(Token.Dot))
            {
                if (!ExpectPeek(Token.Identifier))
                {
                    // TODO : raise error
                    return null;
                }

                instance_get_expression.Index = new StringLiteral { Data = CurrentTokenData };
                
                return instance_get_expression;
            }

            if (CurrentTokenIs(Token.LeftBracket))
            {
                NextToken();
                instance_get_expression.Index = ParseExpression();
                
                if (!ExpectPeek(Token.RightBracket))
                {
                    // TODO : raise error
                    return null;
                }

                return instance_get_expression;
            }

            // TODO : raise error
            return null;
        }

        private Expression ParseReturnExpression()
        {
            ReturnExpression return_expression = new ReturnExpression { Data = CurrentTokenData };

            if (PeekTokenData.Line == CurrentTokenData.Line)
            {
                NextToken();
                return_expression.Value = ParseExpression();
            }

            return return_expression;
        }

        private Expression ParseCallExpression(Expression function)
        {
            CallExpression call_expression = new CallExpression { Function = function };

            NextToken();

            call_expression.Parameters = ParseCommaExpressions(Token.RightParentheses);
            
            return call_expression;
        }

        private Expression ParseLocalExpression()
        {
            if (!ExpectPeek(Token.Identifier))
            {
                // TODO : raise error
            }
            
            LocalExpression local_expression = new LocalExpression { Identifier = (Identifier)ParseIdentifier() };

            if (PeekTokenIs(Token.Assign))
            {
                NextToken();
                NextToken();
                local_expression.Value = ParseExpression();
            }

            return local_expression;
        }

        private bool ExpectToken(Token token_to_check, Token token_expected)
        {
            if (token_to_check == token_expected)
                return true;
            
            // TODO : raise error
            return false;
        }

        private bool ExpectPeek(Token token_expected)
        {
            if (ExpectToken(PeekTokenData.Token, token_expected))
            {
                NextToken();
                return true;
            }

            return false;
        }

        private void NextToken()
        {
            CurrentTokenData = PeekTokenData;

            PeekTokenData = lexer.Lex();
        }

        private bool CurrentTokenIs(Token token)
        {
            return CurrentTokenData.Token == token;
        }

        private bool PeekTokenIs(Token token)
        {
            return PeekTokenData.Token == token;
        }

        private SymbolPriority CurrentPrecedence
        {
            get
            {
                if (precedences.ContainsKey(CurrentTokenData.Token))
                    return precedences[CurrentTokenData.Token];
                return SymbolPriority.Lowest;
            }
        }
        
        private SymbolPriority PeekPrecedence
        {
            get
            {
                if (precedences.ContainsKey(PeekTokenData.Token))
                    return precedences[PeekTokenData.Token];
                return SymbolPriority.Lowest;
            }
        }
        
        public Node Parse(string filename, string source = "")
        {
            if (String.IsNullOrEmpty(source))
                source = file_loader.LoadFile("", filename);
            
            lexer = new Lexer(filename, source);
            
            NextToken();
            NextToken();
            
            return ParseProgram(new[]{ Token.Eof });
        }

        private Program ParseProgram(IEnumerable<Token> terminators)
        {
            Program program = new Program();

            while (true)
            {
                if (terminators.Any(CurrentTokenIs))
                {
                    return program;
                }

                Expression expression = ParseExpression();
                
                if (expression != null)
                    program.Expressions.Add(expression);
                
                NextToken();
            }
        }

        private Expression ParseExpression(SymbolPriority precedence = SymbolPriority.Lowest)
        {
            if (!prefix_functions.ContainsKey(CurrentTokenData.Token))
            {
                // TODO : raise error, prefix function not exists
                throw new Exception("prefix function not exists");
            }
            
            PrefixFunction prefix_function = prefix_functions[CurrentTokenData.Token];

            Expression left_expression = prefix_function();

            while (!PeekTokenIs(Token.Eof) && precedence < PeekPrecedence)
            {
                if (!infix_functions.ContainsKey(PeekTokenData.Token))
                    return left_expression;
                
                NextToken();

                left_expression = infix_functions[CurrentTokenData.Token](left_expression);
            }

            return left_expression;
        }

        private Expression ParseIdentifier()
        {
            return new Identifier { Data = CurrentTokenData };
        }

        private Literal ParseKeywordLiteral()
        {
            if (CurrentTokenIs(Token.Base))
                return new BaseLiteral { Data = CurrentTokenData };

            if (CurrentTokenIs(Token.This))
                return new ThisLiteral { Data = CurrentTokenData };
            
            return new NullLiteral { Data = CurrentTokenData };
        }

        private Literal ParseStringLiteral()
        {
            return new StringLiteral { Data = CurrentTokenData };
        }

        
        private IntegerLiteral ParseIntegerLiteral()
        {
            IntegerLiteral integer_literal = new IntegerLiteral { Data = CurrentTokenData };
            
            if (!Int64.TryParse(integer_literal.Data.Value, out integer_literal.Value))
            {
                // TODO : raise parse error
            }

            return integer_literal;
        }

        private FloatLiteral ParseFloatLiteral()
        {
            FloatLiteral float_literal = new FloatLiteral { Data = CurrentTokenData };

            if (!double.TryParse(float_literal.Data.Value, out float_literal.Value))
            {
                // TODO : raise parse error
            }

            return float_literal;
        }

        private BooleanLiteral ParseBooleanLiteral()
        {
            BooleanLiteral boolean_literal = new BooleanLiteral { Data = CurrentTokenData };

            boolean_literal.Value = CurrentTokenIs(Token.True);

            return boolean_literal;
        }

        private PrefixExpression ParsePrefixExpression()
        {
            PrefixExpression prefix_expression = new PrefixExpression { Data = CurrentTokenData };
            NextToken();
            prefix_expression.Right = ParseExpression(SymbolPriority.Prefix);
            return prefix_expression;
        }

        private Expression ParseGroupExpression()
        {
            NextToken();

            Expression expression = ParseExpression();

            ExpectPeek(Token.RightParentheses);

            return expression;
        }

        private Expression ParseInfixExpression(Expression left)
        {
            InfixExpression infix_expression = new InfixExpression { Left = left, Data = CurrentTokenData };
            SymbolPriority current_precedence = CurrentPrecedence;
            NextToken();
            infix_expression.Right = ParseExpression(current_precedence);
            return infix_expression;
        }

        private Expression ParseIfExpression()
        {
            if (!ExpectPeek(Token.LeftParentheses))
            {
                // TODO : raise error
            }
            
            NextToken();
            
            IfExpression if_expression = new IfExpression { Condition = ParseExpression()};
            
            if (!ExpectPeek(Token.RightParentheses))
            {
                // TODO : raise error
            }
            
            NextToken();

            if_expression.TruePart = ParseProgram(new[]{ Token.Else, Token.End });

            if (CurrentTokenIs(Token.Else))
            {
                NextToken();
                if_expression.FalsePart = ParseProgram(new[]{ Token.End });
            }

            if (!CurrentTokenIs(Token.End))
            {
                // TODO : raise error
            }

            return if_expression;
        }

        private List<LocalExpression> ParseParameters(Token terminator)
        {
            List<LocalExpression> parameters = new List<LocalExpression>();

            bool last_is_comma = false;
            
            while (!(CurrentTokenIs(terminator) || CurrentTokenIs(Token.Eof)))
            {
                if (!CurrentTokenIs(Token.Identifier))
                {
                    // TODO : raise error
                }

                LocalExpression local_expression = new LocalExpression { Identifier = (Identifier)ParseIdentifier() };

                NextToken();

                if (CurrentTokenIs(Token.Assign))
                {
                    NextToken();
                    local_expression.Value = ParseExpression();
                    NextToken();
                }

                parameters.Add(local_expression);
                
                if (CurrentTokenIs(Token.Comma))
                {
                    NextToken();
                    last_is_comma = true;
                }
                else
                {
                    last_is_comma = false;
                }
            }

            if (last_is_comma)
            {
                // TODO : raise error
            }

            return parameters;
        }

        private Expression ParseFunctionExpression()
        {
            if (!ExpectPeek(Token.LeftParentheses))
            {
                // TODO : raise error
            }
            
            NextToken();
            
            FunctionExpression function_expression = new FunctionExpression();

            function_expression.Parameters = ParseParameters(Token.RightParentheses);
            
            if (!CurrentTokenIs(Token.RightParentheses))
            {
                // TODO : raise error
            }
            
            NextToken();

            function_expression.Body = ParseProgram(new[] { Token.End });

            if (!CurrentTokenIs(Token.End))
            {
                // TODO : raise error
            }

            return function_expression;
        }

    }
}